openshift/oc-mirror
# Clid 205

# Description

Before this PR, for release image mirroring, and for the same destination mirror (`mymirror.mydomain`), and same release (`4.15.4-x86_64`) v1 and v2 were using different locations:

- in v1:
  - a release component such as `prometheus-operator`, image `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:eaa2e310dffb4d51375196760a10690c853bf686af7852f61d1878872e058163` was mirrored to `mymirror.mydomain/openshift/release:4.15.4-x86_64-prometheus-operator`
  - a release image `quay.io/openshift-release-dev/ocp-release:4.15.4-x86_64` was mirrored to `mymirror.mydomain/openshift/release-images:4.15.4-x86_64`
- in v2:
  - a release component such as `prometheus-operator`, image `quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:eaa2e310dffb4d51375196760a10690c853bf686af7852f61d1878872e058163` was mirrored to `mymirror.mydomain/openshift-release-dev/ocp-v4.0-art-dev:sha256-eaa2e310dffb4d51375196760a10690c853bf686af7852f61d1878872e058163`
  - a release image `quay.io/openshift-release-dev/ocp-release:4.15.4-x86_64` was mirrored to `mymirror.mydomain/openshift-release-dev/ocp-release:4.15.4-x86_64`

As a result migration from v1 to v2 might be complex, as:
- when mirroring with v2 a certain release, the mirror registry would need to store the images to a different location
- when using the v2 delete feature, a user would not be able to delete release images previously mirrored with v1
- v2's IDMS and v1's ICSP would contain different mirrors, and that might confuse users, or cause problems on the cluster.

With this PR, we reuse the same strategy for the destination of release images as v1, so that the migration is easier.

Please also note that the IDMS generation has been tweeked in order to contain the mirror for the release image (although mirrored by tag) as well as the release component images (mirrored by digest). This is in order to make the change seemless: in fact, users were taking the content of the ICSP and copying it into the install-config.yaml to create a disconnected cluster. And by adding this tweek to the IDMS generation, we are ensuring that the behavior stays almost the same.


Fixes # [CLID-205](https://issues.redhat.com/browse/CLID-205)

## Type of change

Please delete options that are not relevant.


# How Has This Been Tested?

- Improvement of unit tests for v2/internal/pkg/release, in order to assert images returned
- update of Makefile in order to run the v2 end to end tests
- Mirror 2 Mirror
  - Validated that the IDMS, ITMS, UpdateService files generated are as expected
- Mirror 2 Mirror dry-run : and verified that the images included correspond to the ones in the release's image-references file
- Mirror 2 Disk + Disk 2 Mirror 
  - Validated that the IDMS, ITMS, UpdateService files generated are as expected
-  Mirror 2 Disk dry-run
  - Validated that the images included correspond to the ones in the release's image-references 
- Disk 2 Mirror dry-run
  - Validated that the images included correspond to the ones in the release's image-references 
- Delete generate + Delete for same release as the one mirrored
  - verified that the images get deleted
- Run a v1 Mirror 2 mirror, then use delete generate and delete of v2 to delete the release mirrored with v1
  - verified that the images are deleted
